cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/MetricsMessageAggregator.java

```diff
@@ -0,0 +1,297 @@
+      decodingInputStream.reset(message.getPayload());
+      MetricValues values;
+      try {
+        values = reader.read(decoder, schema);
+      } catch (IOException e) {
+        LOG.warn("Failed to decode metric message.", e);
+        return;
+      }
+
+      addToAggregatedValues(aggregated, offset, values);
+    });
+
+    return aggregated.entrySet().stream()
+        .map(this::getMetricValues)
+        .map(metricValues -> encodeMetricValues(metricValues, messageId[0]))
+        .iterator();
+  }
+
+  private void addToAggregatedValues(Map<MetricsMapKey, MetricsAggregator> aggregated,
+      long offset, MetricValues values) {
+    long timestamp = values.getTimestamp();
+    long roundedTimestamp = getRoundedTimestamp(timestamp, offset);
+    MetricsAggregator metricsAggregator = aggregated.computeIfAbsent(
+        new MetricsMapKey(values.getTags(), roundedTimestamp),
+        t -> new MetricsAggregator());
+
+    for (MetricValue metric : values.getMetrics()) {
+      metricsAggregator.addMetric(metric, timestamp);
+    }
+  }
+
+  private MetricValues getMetricValues(
+      Entry<MetricsMapKey, MetricsAggregator> entry) {
+    return entry.getValue().getAggregateMetricValues(
+        entry.getKey().getTags(), entry.getKey().getRoundedTimestamp());
+  }
+
+  private Message encodeMetricValues(MetricValues metricValues,
+      String lastMessageId) {
+    encodeOutputStream.reset();
+    try {
+      recordWriter.encode(metricValues, encoder);
+    } catch (IOException e) {
+      throw new RuntimeException(
+          "Failed to encode metric to message payload.", e);
+    }
+    byte[] payload = encodeOutputStream.toByteArray();
+    return new Message() {
+      @Override
+      public String getId() {
+        // TopicRelayers keep track of the last message ID
+        // to fetch messages in the next iteration.
+        // Message IDs are generated by the messaging service.
+        // Return the last message ID before aggregation every time.
+        return lastMessageId;
+      }
+
+      @Override
+      public byte[] getPayload() {
+        return payload;
+      }
+    };
+  }
+
+  private long getRoundedTimestamp(long timestamp, long offset) {
+    return (Math.max(timestamp - offset, 0) / aggregationWindowSeconds)
+        * aggregationWindowSeconds;
+  }
+
+  /**
+   * The key for aggregating {@link MetricValues}. Only values with the same key
+   * will be merged.
+   */
+  private static class MetricsMapKey {
+
+    private final Map<String, String> tags;
+    private final long roundedTimestamp;
+
+    private MetricsMapKey(Map<String, String> tags, long roundedTimestamp) {
+      this.tags = tags;
+      this.roundedTimestamp = roundedTimestamp;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (!(o instanceof MetricsMapKey)) {
+        return false;
+      }
+      MetricsMapKey that = (MetricsMapKey) o;
```

Comment:
nit: prefer input or something similar compare to `that`

Reply:
Changed the variable name from `that` to `input`. I think this method was generated by IntelliJ.